cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+Each scalar type has a corresponding non-owning device view class which allows access to the value 
+and its validity from the device. This can be obtained using the function 
+`get_scalar_device_view(ScalarType s)`. Note that a device view is not provided for a base scalar 
+object, only for the derived typed scalar class objects.
+
+# libcudf++ API and Implementation
+
+## Streams
+
+CUDA streams are not yet exposed in external libcudf APIs. However, in order to ease the transition 
+to future use of streams, all libcudf APIs that allocate device memory or execute a kernel should be 
+implemented using asynchronous APIs on the default stream (e.g., stream 0).
+
+The recommended pattern for doing this is to make the definition of the external API invoke an 
+internal API in the `detail` namespace. The internal `detail` API has the same parameters as the 
+public API, plus a `rmm::cuda_stream_view` parameter at the end defaulted to 
+`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API 
+definition and use only asynchronous versions of CUDA APIs with the stream parameter.
+
+In order to make the `detail` API callable from other libcudf functions, it should be exposed in a 
+header placed in the `cudf/cpp/include/detail/` directory.
+
+For example:
+
+```c++
+// cpp/include/cudf/header.hpp
+void external_function(...);
+
+// cpp/include/cudf/detail/header.hpp
+namespace detail{
+void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
+} // namespace detail
+
+// cudf/src/implementation.cpp
+namespace detail{
+    // defaulted stream parameter
+    void external_function(..., rmm::cuda_stream_view stream){
+        // implementation uses stream w/ async APIs
+        rmm::device_buffer buff(...,stream);
+        CUDA_TRY(cudaMemcpyAsync(...,stream.value()));
+        kernel<<<..., stream>>>(...);
+        thrust::algorithm(rmm::exec_policy(stream), ...);
+    }
+} // namespace detail
+
+void external_function(...){
+    detail::external_function(...);
+}
+```
+
+**Note:** It's important to synchronize the stream if *and only if* it is necessary. For example,
+when a non-pointer value is returned from the API that is the result of an asynchronous 
+device-to-host copy, the stream used for the copy should be synchronized before returning. However,
+when a column is returned, the stream should not be synchronized because doing so will break 
+asynchrony if and when we add an asynchronous API to libcudf.
+
+**Note:** `cudaDeviceSynchronize()` should *never* be used.
+ This limits the ability to do any multi-stream/multi-threaded work with libcudf APIs.
+
+ ### Stream Creation
+
+There may be times in implementing libcudf features where it would be advantageous to use streams 
+*internally*, i.e., to accomplish overlap in implementing an algorithm. However, dynamically 
+creating a stream can be expensive. RMM has recently added a stream pool for this situation to 
```

Comment:
```suggestion
creating a stream can be expensive. RMM has a stream pool for this situation to 
```

Reply:
Hmm... Please feel free to ignore this suggestion. I wanted to avoid references to points in time, to prevent said reference from going out of date. But I see that the following sentence refers to this as `not yet exposed in libcudf`.

Reply:
I generally agree with your sentiment here. I'll improve it.